search-replace.go

```diff
@@ -155,71 +162,229 @@ func fixLine(line *[]byte, replacements []*Replacement) *[]byte {
 
-		// Update the serialized content inline
+			// since we've filtered out all the escaped value already, this should be the actual end
+			nextPartIndex = currentContentIndex + 3
+			details.NextPartIndex = nextPartIndex
+			// we're at backslash, so we need to minus 1 to get the index where the content finishes
+			details.ContentEndIndex = currentContentIndex - 1
+			nextPartFound = true
+			break
+		}
+
+		contentByteCount++
+		currentContentIndex++
+	}
+
+	if nextPartFound == false {
+		return nil, fmt.Errorf("end of serialized string not found")
+	}
+
+	return &details, nil
+}
 
-		//contentEnd = contentStart + newLength // adjust the end index based on the new length -- THIS BREAKS THINGS
+func getUnescapedBytesIfEscaped(charPair []byte) []byte {
 
-		*line = append((*line)[:contentStart], append(updatedContent, (*line)[contentEnd:]...)...)
-		Debugf("After updating content: %s\n", string(*line))
+	backslash := byte('\\')
 
-		// Adjust startIndex for the next iteration
-		startIndex += match[1] + newLength + len(newLengthStr) - len((*line)[startIndex+match[2]:startIndex+match[3]])
-		Debugf("New startIndex: %d\n", startIndex)
+	//escapables := []byte{'\\', '\'', '"', 'n', 'r', 't', 'b', 'f', '0'}
+
+	// a map of the second byte to its actual binary presentation
+
+	// if the first byte is not a backslash, we don't need to do anything
+
+	if charPair[0] != backslash {
+		return charPair
 	}
 
-	return line
+	unescapedMap := map[byte]byte{
+		'\\': '\\',
+		'\'': '\'',
+		'"':  '"',
+		'n':  '\n',
+		'r':  '\r',
+		't':  '\t',
+		'b':  '\b',
+		'f':  '\f',
+		'0':  '\x00',
+	}
+
+	actualByte := unescapedMap[charPair[1]]
+
+	if actualByte != 0 {
+		return []byte{actualByte}
+	}
+
+	// what if it's not a valid escape? Do nothing - it's considered as already escaped
+	return charPair
 }
 
-func replaceInSerializedBytes(serialized []byte, replacements []*Replacement) []byte {
-	for _, replacement := range replacements {
-		serialized = bytes.ReplaceAll(serialized, replacement.From, replacement.To)
+func unescapeContent(escaped []byte) []byte {
+	unescapedBytes := make([]byte, 0, len(escaped))
+	index := 0
+
+	// only applies to content - do not apply to raw mysql query
+	// tested with php -i, mysql client, and mysqldump and mydumper.
+	// 1. \" in dump becomes " when inserting a mysql row.
+	// 2. \\ in dump becomes \ when inserting a mysql row.
+	// 3. \' in dump becomes ' when inserting a mysql row.
+	// 4. mysql translates newline into \n when creating a mysqldump. Same applies to carriage return.
+	// 5. PHP serialize does not convert the bytes \r or \n into something else - they're as-is.
+	// 6. If using single quotes in php, \r and \n does not get converted into bytes - they become literal backslash and letter.
+	// Generally, to unescape, we need to do the following:
+	// 1. Convert \\ to \
+	// 2. Convert \' to '
+	// 3. Convert \" to "
+
```

Comment:
I'll add a few more conditions to this PR here - we should be converting everything within `unescapedMap` just above. Not sure what else I missed.